### PR TITLE
Add test compilation unit component for integration test targets

### PR DESCRIPTION
**Pull request by maciektr**
commit-id:ae5578cd

---

**Stack**:
- #688
- #679
- #651
- #649
- #628
- #660
- #615
- #597 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: scarb/src/ops/resolve.rs [hunk 1]
```diff
@@ -222,6 +265,13 @@ fn generate_cairo_compilation_units(
         .collect())
 }
 
+fn is_integration_test(target: &Target, lib_target: Option<Target>) -> bool {
+    let lib_target_source_path = lib_target
+        .map(|target| target.source_path.to_string())
+        .unwrap_or(DEFAULT_SOURCE_PATH.to_string());
+    target.is_test() && !target.source_path.ends_with(lib_target_source_path)
```
Comment (maciektr) [inline scarb/src/ops/resolve.rs:272]: Not sure about this cond 🤔

File: scarb/src/core/package/id.rs [hunk 2]
```diff
@@ -9,6 +9,7 @@ use crate::core::source::SourceId;
 use crate::core::PackageName;
 use crate::internal::static_hash_cache::StaticHashCache;
 use crate::internal::to_version::ToVersion;
+const TEST_PACKAGE_PREFIX: &str = "___test_package_prefix___";
```
Comment (mkaput) [inline scarb/src/core/package/id.rs:12]: Out of curiosity, how does Cargo prefix these? I'm afraid we'll end up showing this prefix to users for example in error messages, so maybe we should come up with something more "readable". Say `X_unittests`

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/core/manifest/summary.rs [hunk 3]
```diff
@@ -61,10 +63,22 @@ impl Summary {
...
+                .source_id(SourceId::default())
+                .version_req(DependencyVersionReq::exact(&cairo_version))
+                .build()
+        });
+
         let mut deps: Vec<&ManifestDependency> = Vec::new();
 
         if !self.no_core {
             deps.push(&CORE_DEPENDENCY);
+            deps.push(&TEST_PLUGIN_DEPENDENCY);
```
[the remark this answers is not in the document]
Comment (maciektr) [reply scarb/src/core/manifest/summary.rs:81]: Isn't filtering based on DepKind enough? 🤔

File: scarb/Cargo.toml [hunk 4]
```diff
@@ -24,6 +24,7 @@ cairo-lang-semantic.workspace = true
 cairo-lang-sierra-to-casm.workspace = true
 cairo-lang-sierra.workspace = true
 cairo-lang-starknet.workspace = true
+cairo-lang-test-runner.workspace = true
```
[the remark this answers is not in the document]
Comment (maciektr) [reply scarb/Cargo.toml:27]: https://github.com/software-mansion/scarb/issues/652

File: scarb/src/core/manifest/summary.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkaput) [reply scarb/src/core/manifest/summary.rs:81]: Not sure I follow, can you elaborate?

[the remark this answers is not in the document]
Comment (maciektr) [reply scarb/src/core/manifest/summary.rs:81]: I just thought that we can add this test plugin dependency to all packages for simplicity, as it is only pulled for test targets during resolution anyway.

[the remark this answers is not in the document]
Comment (mkaput) [reply scarb/src/core/manifest/summary.rs:81]: trick is that this method would end up polluting people's lockfile which is something I'd rather avoid

File: scarb/src/ops/resolve.rs (hunk 1 above)
Comment (mkaput) [reply scarb/src/ops/resolve.rs:280]: Either this stays, it's quite fine, or we store an enum in test target props which states whether this is unit test or integration test.

File: scarb/src/core/package/id.rs [hunk 5]
```diff
@@ -136,7 +142,8 @@ impl<'de> Deserialize<'de> for PackageId {
 
 impl fmt::Display for PackageId {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{} v{}", self.name, self.version)?;
+        let name = self.name.to_string().replace(TEST_PACKAGE_PREFIX, "");
+        write!(f, "{} v{}", name, self.version)?;
```
Comment (mkaput) [inline scarb/src/core/package/id.rs:146]: I have mixed feelings on this. This loses potentially important information. Can we drop this change?

**Review by mkaput — CHANGES_REQUESTED** [review]

**Review by mkaput — APPROVED** [review]

**Pull request merged**